### PR TITLE
Fresh prefab

### PR DESCRIPTION
**Pull request by mmore500**
It's PR #290 but cherry-picked to branch off of master, not memic_merge

### CODE REVIEW

**Comment by mmore500** [comment]
- [x] rebase so it's not off of memic_merge branch
- [ ] pass for general C++ style in code files
- [ ] make tests less dependent on specific order of css classes
- [ ] refactor css to reduce duplication (less css?)
- [ ] try spacing clicks out by one second and see if the clicking works in tests

**Comment by mmore500** [comment]
reminder to self: merge https://github.com/devosoft/cookiecutter-empirical-project/pull/30 when merging this!

**Comment by mmore500** [comment]
remaining tasks:
- [ ] make a demo file for clicks
- [ ] @mmore500 do a code review

File: source/prefab/Card.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+          prefab::FontAwesomeIcon down("fa-angle-double-down");
+          card_header << up;
+          card_header << down;
+          up.AddAttr("class", "toggle_glyph");
+          down.AddAttr("class", "toggle_glyph");
+        }
+        card_header.AddAttr("class", "collapse_toggle_card_header");
+      }
+    }
+
```
[the remark this answers is not in the document]
Comment (mmore500) [reply source/prefab/Card.h:64]: oh, referring to the /* * */ instead of /// below

File: tests/web/TestClickTest.cc [hunk 2]
```diff
@@ -0,0 +1,278 @@
+//  This file is part of Empirical, https://github.com/devosoft/Empirical
+//  Copyright (C) Michigan State University, 2020.
+//  Released under the MIT Software license; see doc/LICENSE
+
+/*
+ * This file can be used to test triggering a mouse click by the user.
```
Comment (mmore100) [inline tests/web/TestClickTest.cc:6]: Amazing,!!!

File: tests/web/ClickDemo.cc [hunk 3]
```diff
@@ -0,0 +1,107 @@
+//  This file is part of Empirical, https://github.com/devosoft/Empirical
```
Comment (mmore100) [inline tests/web/ClickDemo.cc:1]: I think this file has been superseded by the other one?

**Review by mmore100 — APPROVED** [review]
A few polishes and this is ready to go. Adds major features to empirical web library, lots more work to do and good detailed descriptions of that work included

File: source/prefab/Card.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (kayakingCellist) [reply source/prefab/Card.h:64]: oh, ok, thanks!

File: source/tools/string_utils.h [hunk 4]
```diff
@@ -287,6 +305,26 @@ namespace emp {
...
+    constexpr int char_shift = 'a' - 'A';
+    bool next_upper = true;
+    for (size_t i = 0; i < value.size(); i++) {
+      if (next_upper && value[i] >= 'a' && value[i] <= 'z') {
+        value[i] = (char) (value[i] - char_shift);
+      } else if (!next_upper && value[i] >= 'A' && value[i] <= 'Z') {
+        value[i] = (char) (value[i] + char_shift);
+      }
+
+      if (value[i] == ' ') {
```
[the remark this answers is not in the document]
Comment (kayakingCellist) [reply source/tools/string_utils.h:319]: I see the efficiency, but I didn't add this function. Do you think the person who did would be ok if we changed it?

File: source/tools/string_utils.h [hunk 5]
```diff
@@ -862,6 +900,24 @@ namespace emp {
     return out_val;
   }
 
+  template <typename T>
+  inline std::string join(const emp::vector<T> & v, std::string join_str) {
```
[the remark this answers is not in the document]
Comment (kayakingCellist) [reply source/tools/string_utils.h:904]: Same as above. I think the function is returning a list of the values of vector `v` in string format separated by  `join_str`. Correct me if I'm wrong. Would it ever be the case that `T` could not be cast to a string? Would that throw an error?

File: source/web/Widget.h [hunk 6]
```diff
@@ -332,7 +336,14 @@ namespace web {
       virtual void GetHTML(std::stringstream & ss) = 0;
 
       // Derived widgets may also provide JavaScript code to be run on redraw.
-      virtual void TriggerJS() { ; }
+      virtual void TriggerJS() {
+        on_update_js_signal.Trigger();
+      }
+
+      // Add JS function to be executed when TriggerJS() is called
+      SignalKey DoUpdateJS(const std::function<void()> &fun){
```
[the remark this answers is not in the document]
Comment (kayakingCellist) [reply source/web/Widget.h:344]: Switched to `RegisterUpdateJS`

File: tests/web/ClickDemo.cc (hunk 3 above)
Comment (kayakingCellist) [reply tests/web/ClickDemo.cc:1]: I think the other one shows our thought process more and this one is a more "clean" example. Do you think we should keep both or choose one?

Comment (mmore500) [reply tests/web/ClickDemo.cc:1]: keeping both!

**Pull request merged**